### PR TITLE
[HCS] Extend to outline overlapping sub/super cold regions

### PR DESCRIPTION
**Pull request by scui-ibm**
Currently, with hot cold splitting, when a cold region is identified, it is added to the region list of ColdBlocks. Then when another cold region (B) identified overlaps with a ColdBlocks region (A) already added to the list, the region B is not added to the list because of the overlapping with region A. The splitting analysis is performed, and the region A may not get split, for example, if it’s considered too expansive. This is to improve the handling the overlapping case when the region A is not considered good for splitting, while the region B is good for splitting.
 
The change is to move the cold region splitting analysis earlier to allow more cold region splitting. If an identified region cannot be split, it will not be added to the candidate list of ColdBlocks for overlapping check.

### CODE REVIEW

**Comment by scui-ibm** [comment]
Gentle ping for review. And if you have any suggestions of other reviewers, please let me know. Thanks!

**Comment by hiraditya** [comment]
I'll review this in a week as I'm out for a few days. Sorry for holding you back.

**Comment by hiraditya** [comment]
In the meantime, do you mind reviewing why windows test failed? Maybe rebasing on top of main will fix it?

Any benchmark or performance numbers you have would also help motivate incorporating your changes.

File: llvm/lib/Transforms/IPO/HotColdSplitting.cpp [hunk 1]
```diff
@@ -394,9 +382,16 @@ Function *HotColdSplitting::extractColdRegion(
...
                     << ", penalty = " << OutliningPenalty << "\n");
   if (!OutliningBenefit.isValid() || OutliningBenefit <= OutliningPenalty)
-    return nullptr;
+    return false;
+
+  return true;
+}
 
-  Function *OrigF = Region[0]->getParent();
+Function *HotColdSplitting::extractColdRegion(
```
Comment (hiraditya) [inline llvm/lib/Transforms/IPO/HotColdSplitting.cpp:390]: Please add comment to clarify how BB is sufficient to extract a region. Is it a single-entry single-exit region?

**Review by hiraditya — APPROVED** [review]
Overall lgtm, please add comments as mentioned in the review. Thanks for working on this.

**Pull request merged**